File: backend/app/services/authorized_pickup_service.py
from app.constants import MAX_AUTHORIZED_PICKUPS, PICKUP_ID_TYPES
from app.extensions import db
from app.models.authorized_pickup import AuthorizedPickupPerson
from app.models.user import User
from app.services.auth_service import normalize_phone


def list_authorized_pickups(customer: User) -> list[AuthorizedPickupPerson]:
    return (
        AuthorizedPickupPerson.query.filter_by(customer_id=customer.id)
        .order_by(AuthorizedPickupPerson.sort_order, AuthorizedPickupPerson.created_at)
        .all()
    )


def count_authorized_pickups(customer: User) -> int:
    return AuthorizedPickupPerson.query.filter_by(customer_id=customer.id).count()


def get_authorized_pickup(customer: User, pickup_id) -> AuthorizedPickupPerson | None:
    return AuthorizedPickupPerson.query.filter_by(
        id=pickup_id, customer_id=customer.id
    ).first()


def _validate_id_type(value: str) -> str:
    value = (value or "").strip().lower()
    if value not in PICKUP_ID_TYPES:
        raise ValueError("Invalid ID type")
    return value


def create_authorized_pickup(customer: User, data: dict) -> AuthorizedPickupPerson:
    if count_authorized_pickups(customer) >= MAX_AUTHORIZED_PICKUPS:
        raise ValueError(f"You can save up to {MAX_AUTHORIZED_PICKUPS} authorized pickup persons")

    full_name = (data.get("full_name") or "").strip()
    if not full_name:
        raise ValueError("Full name is required")

    pickup = AuthorizedPickupPerson(
        customer_id=customer.id,
        full_name=full_name,
        relationship="other",
        contact_number=normalize_phone(data.get("contact_number") or ""),
        id_type=_validate_id_type(data.get("id_type")),
        id_last_four=None,
        notes=(data.get("notes") or "").strip() or None,
        sort_order=count_authorized_pickups(customer),
    )
    db.session.add(pickup)
    db.session.commit()
    return pickup


def update_authorized_pickup(pickup: AuthorizedPickupPerson, data: dict) -> AuthorizedPickupPerson:
    if "full_name" in data:
        full_name = (data.get("full_name") or "").strip()
        if not full_name:
            raise ValueError("Full name cannot be empty")
        pickup.full_name = full_name

    if "contact_number" in data:
        pickup.contact_number = normalize_phone(data.get("contact_number") or "")

    if "id_type" in data:
        pickup.id_type = _validate_id_type(data.get("id_type"))

    if "notes" in data:
        pickup.notes = (data.get("notes") or "").strip() or None

    db.session.commit()
    return pickup


def delete_authorized_pickup(pickup: AuthorizedPickupPerson) -> None:
    db.session.delete(pickup)
    db.session.commit()

File: server/src/services/authorizedPickupService.ts
import { asc, eq, sql } from 'drizzle-orm'
import { MAX_AUTHORIZED_PICKUPS, PICKUP_ID_TYPES } from '../constants.js'
import { db } from '../db/index.js'
import { authorizedPickupPersons, type UserRow } from '../db/schema/index.js'
import { normalizePhone } from './authService.js'

type AuthorizedPickupRow = typeof authorizedPickupPersons.$inferSelect

export async function listAuthorizedPickups(customer: UserRow): Promise<AuthorizedPickupRow[]> {
  return db
    .select()
    .from(authorizedPickupPersons)
    .where(eq(authorizedPickupPersons.customerId, customer.id))
    .orderBy(asc(authorizedPickupPersons.sortOrder), asc(authorizedPickupPersons.createdAt))
}

async function countAuthorizedPickups(customer: UserRow): Promise<number> {
  const rows = await db
    .select({ id: authorizedPickupPersons.id })
    .from(authorizedPickupPersons)
    .where(eq(authorizedPickupPersons.customerId, customer.id))
  return rows.length
}

export async function getAuthorizedPickup(
  customer: UserRow,
  pickupId: string,
): Promise<AuthorizedPickupRow | null> {
  const [row] = await db
    .select()
    .from(authorizedPickupPersons)
    .where(eq(authorizedPickupPersons.id, pickupId))
    .limit(1)
  if (!row || row.customerId !== customer.id) return null
  return row
}

function validateIdType(value: string): string {
  const normalized = (value ?? '').trim().toLowerCase()
  if (!(PICKUP_ID_TYPES as readonly string[]).includes(normalized)) {
    throw new Error('Invalid ID type')
  }
  return normalized
}

export async function createAuthorizedPickup(
  customer: UserRow,
  data: Record<string, unknown>,
): Promise<AuthorizedPickupRow> {
  const count = await countAuthorizedPickups(customer)
  if (count >= MAX_AUTHORIZED_PICKUPS) {
    throw new Error(`You can save up to ${MAX_AUTHORIZED_PICKUPS} authorized pickup persons`)
  }

  const fullName = String(data.full_name ?? '').trim()
  if (!fullName) throw new Error('Full name is required')

  const [pickup] = await db
    .insert(authorizedPickupPersons)
    .values({
      customerId: customer.id,
      fullName,
      relationship: 'other',
      contactNumber: normalizePhone(String(data.contact_number ?? '')),
      idType: validateIdType(String(data.id_type ?? '')),
      notes: String(data.notes ?? '').trim() || null,
      sortOrder: count,
    })
    .returning()
  return pickup!
}

export async function updateAuthorizedPickup(
  pickup: AuthorizedPickupRow,
  data: Record<string, unknown>,
): Promise<AuthorizedPickupRow> {
  const updates: Partial<typeof authorizedPickupPersons.$inferInsert> = {}

  if ('full_name' in data) {
    const fullName = String(data.full_name ?? '').trim()
    if (!fullName) throw new Error('Full name cannot be empty')
    updates.fullName = fullName
  }
  if ('contact_number' in data) {
    updates.contactNumber = normalizePhone(String(data.contact_number ?? ''))
  }
  if ('id_type' in data) {
    updates.idType = validateIdType(String(data.id_type ?? ''))
  }
  if ('notes' in data) {
    updates.notes = String(data.notes ?? '').trim() || null
  }

  updates.updatedAt = new Date()
  const [updated] = await db
    .update(authorizedPickupPersons)
    .set(updates)
    .where(eq(authorizedPickupPersons.id, pickup.id))
    .returning()
  return updated!
}

export async function deleteAuthorizedPickup(pickup: AuthorizedPickupRow): Promise<void> {
  await db.delete(authorizedPickupPersons).where(eq(authorizedPickupPersons.id, pickup.id))
}
